Source/Dafny/Plugins/Rewriter.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System.Diagnostics.Contracts;
+
+namespace Microsoft.Dafny.Plugins {
+  /// <summary>
+  /// A class that plugins should extend, in order to provide an extra Rewriter to the pipeline.
+  ///
+  /// If the plugin defines no Configuration, then Dafny will instantiate every sub-class
+  /// of Rewriter from the plugin, providing them with an ErrorReporter as the first and only argument.
+  /// </summary>
+  public abstract class Rewriter {
```

Comment:
Is there a need for plugins that aim to support custom compiler back-ends to override anything other than the final `PostResolve(Program program)` ?

I feel the use-case of this PR is underspecified. The title reads "This PR creates an extensible plugin structure, which is primarily used to provide Dafny with extra Rewriters", which I think is more of a how than a why.

My impression from our conversations is that this PR is meant to allow custom back-ends to add additional error reporting. I would think that this only requires adding errors after resolution has completed entirely. Is that right?

Reply:
> Is there a need for plugins that aim to support custom compiler back-ends to override anything other than the final PostResolve(Program program) ?

Plugin support is not about just compiler back-ends (see last answer).

> I think is more of a how than a why.

I added a statement after the how to explain why.

> My impression from our conversations is that this PR is meant to allow custom back-ends to add additional error reporting. I would think that this only requires adding errors after resolution has completed entirely. Is that right?

This PR goes beyond our conversation indeed. It opens the door to plugins actually rewriting the AST before it is resolved (see PreResolve()), and with little effort to later support custom back-ends later.
Hence it's not just "custom back-ends", since we could definitely have plugins that do other checks that are not part of any back-end. For example, a plugin that verifies some naming conventions (think of a linter), another that extern methods are tested, another plugin performs some tests on external methods,, another plugin that replaces some macros before resolution... support custom back-ends.